core/msg.c

```diff
@@ -187,41 +188,90 @@ int msg_send_to_self(msg_t *m)
     return res;
 }
 
-int msg_send_int(msg_t *m, kernel_pid_t target_pid)
+static int _msg_send_oneway(msg_t *m, kernel_pid_t target_pid, bool *sched)
 {
 #ifdef DEVELHELP
     if (!pid_is_valid(target_pid)) {
-        DEBUG("msg_send(): target_pid is invalid, continuing anyways\n");
+        DEBUG("%s: target_pid is invalid, continuing anyways\n", __func__);
     }
 #endif /* DEVELHELP */
 
     thread_t *target = (thread_t *)sched_threads[target_pid];
 
     if (target == NULL) {
-        DEBUG("msg_send_int(): target thread does not exist\n");
+        DEBUG("%s: target thread %d does not exist\n", __func__, target_pid);
         return -1;
     }
 
-    m->sender_pid = KERNEL_PID_ISR;
     if (target->status == STATUS_RECEIVE_BLOCKED) {
-        DEBUG("msg_send_int: Direct msg copy from %" PRIkernel_pid " to %"
-              PRIkernel_pid ".\n", thread_getpid(), target_pid);
-
+        DEBUG("%s: Direct msg copy from %" PRIkernel_pid " to %"
+              PRIkernel_pid ".\n", __func__, thread_getpid(), target_pid);
 
         /* copy msg to target */
         msg_t *target_message = (msg_t *)target->wait_data;
         *target_message = *m;
+
         sched_set_status(target, STATUS_PENDING);
+        *sched = true;
 
-        sched_context_switch_request = 1;
         return 1;
     }
     else {
-        DEBUG("msg_send_int: Receiver not waiting.\n");
+        DEBUG("%s: Receiver not waiting.\n", __func__);
         return (queue_msg(target, m));
     }
 }
 
+int msg_send_int(msg_t *m, kernel_pid_t target_pid)
+{
+    int res;
+    bool sched = false;
+
+    m->sender_pid = KERNEL_PID_ISR;
+
+    res = _msg_send_oneway(m, target_pid, &sched);
+
+    if (sched) {
+        sched_context_switch_request = 1;
+    }
+
+    return res;
+}
+
+int msg_send_bus(msg_t *m, msb_bus_t *bus)
+{
+    const bool in_irq = irq_is_in();
+    const uint32_t event_mask = (1 << (m->type & 0x1F));
+    int count = 0;
+    bool sched = false;
+
+    m->sender_pid = in_irq ? KERNEL_PID_ISR : sched_active_pid;
+
+    unsigned state = irq_disable();
+
+    for (list_node_t *e = bus->subs.next; e; e = e->next) {
+        msg_bus_entry_t *subscriber = container_of(e, msg_bus_entry_t, next);
+
+        if ((subscriber->event_mask & event_mask) == 0) {
+            continue;
+        }
+
+        if (_msg_send_oneway(m, subscriber->pid, &sched) > 0) {
+            ++count;
+        }
+    }
+
+    irq_restore(state);
+
+    if (sched && in_irq) {
```

Comment:
when re-using `sched_context_switch_request` in `_msg_send_oneway()`, this can be simplified to `if (!in_irq && sched_context_switch_request) thread_yield_higher();`.

Reply:
Is it valid to set `sched_context_switch_request` when not in interrupt context?

Reply:
it is valid to set it, not valid to un-set it.

Reply:
And it's valid because it's called while interrupts are disabled, so no sched event can happen in between that would unset it.